pkg/virtctl/expose/expose.go

```diff
@@ -0,0 +1,211 @@
+	cmd.Flags().StringVar(&loadBalancerIP, "load-balancer-ip", "", "IP to assign to the Load Balancer. If empty, an ephemeral IP will be created and used.")
+	cmd.Flags().Int32Var(&port, "port", 0, "The port that the service should serve on")
+	cmd.MarkFlagRequired("port")
+	cmd.Flags().StringVar(&strProtocol, "protocol", "TCP", "The network protocol for the service to be created.")
+	cmd.Flags().IntVar(&intTargetPort, "target-port", 0, "Name or number for the port on the VM that the service should direct traffic to. Optional.")
+	cmd.Flags().Int32Var(&nodePort, "node-port", 0, "Port used to expose the service on each node in a cluster.")
+	cmd.Flags().StringVar(&strServiceType, "type", "ClusterIP", "Type for this service: ClusterIP, NodePort, or LoadBalancer.")
+	cmd.Flags().StringVar(&portName, "port-name", "", "Name of the port. Optional.")
+	cmd.SetUsageTemplate(templates.UsageTemplate())
+
+	return cmd
+}
+
+func usage() string {
+	usage := `  # Expose SSH to a virtual machine called 'myvm' as a node port (5555) of the cluster:
+  virtctl expose vm myvm --port=5555 --target-port=22 --name=myvm-ssh --type=NodePort")`
+	return usage
+}
+
+// executing the "expose" command
+func (o *Command) RunE(cmd *cobra.Command, args []string) error {
+	// first argument is type of VM: VM, offline VM or replica set VM
+	vmType := strings.ToLower(args[0])
+	// second argument must be name of the VM
+	vmName := args[1]
+
+	// these are used to convert the flag values into service spec values
+	var protocol v1.Protocol
+	var targetPort intstr.IntOrString
+	var serviceType v1.ServiceType
+
+	// convert from integer to the IntOrString type
+	targetPort = intstr.FromInt(intTargetPort)
+
+	// convert from string to the protocol enum
+	switch strProtocol {
+	case "TCP":
+		protocol = v1.ProtocolTCP
+	case "UDP":
+		protocol = v1.ProtocolUDP
+	default:
+		return fmt.Errorf("unknown protocol: %s", strProtocol)
+	}
+
+	// convert from string to the service type enum
+	switch strServiceType {
+	case "ClusterIP":
+		serviceType = v1.ServiceTypeClusterIP
+	case "NodePort":
+		serviceType = v1.ServiceTypeNodePort
+	case "LoadBalancer":
+		serviceType = v1.ServiceTypeLoadBalancer
+	case "ExternalName":
+		return fmt.Errorf("type: %s not supported", strServiceType)
+	default:
+		return fmt.Errorf("unknown service type: %s", strServiceType)
+	}
+
+	// get the namespace
+	namespace, _, err := o.clientConfig.Namespace()
+	if err != nil {
+		return err
+	}
+
+	// get the client
+	virtClient, err := kubecli.GetKubevirtClientFromClientConfig(o.clientConfig)
+	if err != nil {
+		return fmt.Errorf("cannot obtain KubeVirt client: %v", err)
+	}
+
+	// does a plain quorum read from the apiserver
+	options := k8smetav1.GetOptions{}
+
+	switch vmType {
+	case "vm", "vms", "virtualmachine", "virtualmachines":
+		// get the VM
+		_, err := virtClient.VM(namespace).Get(vmName, options)
```

Comment:
In this case you have to use `spec.Labels` from the vm.

Also we should probably ignore all `kubevirt.io` labels. They are used for internal orchestration only now  and the expose result would differ from an expose on a pod.

Reply:
Hm, thinking again about it, we have already a special label `kubevirt.i/os` which we definitely do not want to include.